## generator简介

我们在上一节[迭代器](/pages/4b282c/)中已经简单介绍了生成器的基本使用，它能够用一个 `*` 号 于 `yield` 语句帮我们快速生成我们所需的迭代器对象，在本节中，我们会详细讨论生成器的剩余内容，并解析一个较为复杂的例子：如何用生成器构造一个无限的素数序列。

<<<@/codes/javaScript/es6/generator/index.js

生成器是一个*惰性序列*，**用到其中的值**才会执行生成器函数。带着这个思路或许更容易理解在for循环中生成器是如何工作的。

例如例子中的 `for(const item of range(1,4))` 
- 访问了 `item` ，所以进入 `range`执行，执行到 `yield` 停止
- 直到下一轮循环开始，继续访问`item` ，恢复执行，执行到下一个`yield`停止

## return 

我们已经知道，用for循环去遍历一个生成器，获取到的值将是生成器函数每次yield出的值。而生成器既然是一个函数，它自然具有 `return` 语句。我们先来看一下有return 语句时，生成器会如何工作：

<<< @/codes/javaScript/es6/generator/return.js

为了了解 `yield` 和 `return` 的区别，我们需要先了解一下generator是如何工作的：

-   当我们遍历一个生成器对象或对他调用`next`时（第一轮循环），它会执行到第一个yield语句，以上面的例子来说就是执行到 `yield hello` ，然后，生成器函数会 **冻结**。也就是说，打印语句 `console.log('hello')` 不会被执行
-   当进入第二轮循环时，生成器函数会**解冻**，向下执行到下一个yield语句并**再次冻结**。即打印`console.log('hello')`并 `yield world`
-   第三轮循环，继续向下执行，遇到 `return` 语句，循环结束。

## yield

我们刚才的例子都是将 `yield` 单纯作为抛出语句来运用，事实上，**yield是具有返回值的**。这需要我们直接调用 `next()` 语句来执行生成器函数。给next函数传入的参数（空就是undefined）会作为yield语句的返回值，在下一轮执行时（函数解冻时）被获取。

<<< @/codes/javaScript/es6/generator/yield.js#yield1

## throw

`Generator.prototype.throw()` 向生成器抛出一个错误

我们可以在生成器函数的内部部署 `try catch` 语句，在调用生成器函数时通过 `生成器.throw()` 向其抛出错误并在内部进行捕获。

<<< @/codes/javaScript/es6/generator/throw.js

需要注意的地方有两点：
1.  我们刚才说过生成器函数执行到`yield`语句后会被冻结，在这里例子中更为明显。在调用`it.throw()`时，函数刚刚被解冻，**并没有离开 try，catch 范围**，因此会被生成器内部的 `try,catch`捕获错误。
2.  被生成器内部的错误捕获语句捕获后，生成器函数会**继续执行**，直到获取到下一个`yield`语句抛出的值


## yield*

`yield*` 意思是抛出一个新的生成器。我们假设有一个列表生成器，每次抛出一个值，这个值需要第二个生成器进行处理：

<<< @codes/javaScript/es6/generator/nest.js

详细执行流程如下：

![](https://linyc.oss-cn-beijing.aliyuncs.com/20220531140351.png)

这种嵌套结构总是让人挠头，其实你完全可以将结构**拍平**（它只是个数组）,只看普通的yield，并把其排成数组顺序执行：

![](https://linyc.oss-cn-beijing.aliyuncs.com/20220531140812.png)

